---
title: "Uniform Interface for Genomic Variation Signature Analysis and Visualization"
author: "Shixiang Wang \\

        ShanghaiTech. University"
date: "`r Sys.Date()`"
output:
  prettydoc::html_pretty:
    toc: true
    theme: cayman
    highlight: github
  pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Usage documentation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Overview

The goal of **sigminer** is to provide an uniform interface for genomic variation signature analysis and visualization. **sigminer** is originated from [VSHunter](https://github.com/ShixiangWang/VSHunter) package I wrote. I hate ugly structure and function names in VSHunter, thus reconstruct it using concise function names, S4 classes and S3 methods etc.. I will continue to add more features to uncover genomic variation signatures and their correlationship with phenotypes and genotypes.

**sigminer** is powered by [NMF](https://github.com/renozao/NMF) package and [maftools](https://github.com/PoisonAlien/maftools) package.

## Installation

You can install the stable release of **sigminer** from CRAN with:

```{r, eval=FALSE}
install.packages("sigminer")
```

You can also install the development version of sigminer from Github with:

```{r, eval=FALSE}
remotes::install_github("ShixiangWang/sigminer")
```


## Usage

An example for how to extract mutational signatures are given as the following.

```{r}
library(sigminer)
```

### Load data as a MAF object

```{r}
laml.maf <- system.file("extdata", "tcga_laml.maf.gz", package = "maftools")
laml <- read_maf(maf = laml.maf)
```



### Prepare data for signature analysis

```{r}
library(BSgenome.Hsapiens.UCSC.hg19, quietly = TRUE)

sig_pre <- sig_prepare(laml, ref_genome = "BSgenome.Hsapiens.UCSC.hg19", 
                      prefix = "chr", add = TRUE)
```


### Extract signatures

Before extracting signatures, we can estimate signature number with `sig_estimate` function.

```{r}
library(NMF)
```


```{r, eval=FALSE}
sig_est <- sig_estimate(sig_pre$nmf_matrix, range = 2:5, pConstant = 0.01)
```

`sig_est` keeps all information of estimation.


```{r, eval=FALSE}
sig_laml <- sig_extract(sig_pre$nmf_matrix, n_sig = 2, mode = "mutation", pConstant = 0.01)
```

> Please note the 'pConstant' option: 	
It is a small positive value to add to the matrix. Use it ONLY if the functions throws an non-conformable arrays error.

Plot mutational signatures.

```{r, fig.width=10, eval=FALSE}
draw_sig_profile(sig_laml$nmfObj, mode = "mutation")
```

Plot cosine similarities against validated signatures.

```{r, eval=FALSE}
pheatmap::pheatmap(mat= sig_laml$coSineSimMat, cluster_rows = FALSE, 
                   main = "Cosine similarity against validated COSMIC signatures")
```


## Citation

If you use **sigminer** in academic field, please cite:

```
Wang, Shixiang, et al. “APOBEC3B and APOBEC mutational signature as
    potential predictive markers for immunotherapy response in non-small
    cell lung cancer.” Oncogene (2018).
```

and

```
Gaujoux, Renaud, and Cathal Seoighe. "A Flexible R Package for 
    Nonnegative Matrix Factorization."" BMC Bioinformatics 11, no. 1 (December 2010).
```

## Acknowledgments

The code for extracting copy number signatures was based in part on the source code 
from paper *Copy number signatures and mutational processes in ovarian carcinoma*, 
if you use this feature, please also cite:

```
Macintyre, Geoff, et al. "Copy number signatures and mutational
    processes in ovarian carcinoma." Nature genetics 50.9 (2018): 1262.
```

The code for extracting mutational signatures was based in part on the source 
code of the **maftools** package, if you use this feature, please also cite:

```
Mayakonda, Anand, et al. "Maftools: efficient and comprehensive analysis
    of somatic variants in cancer." Genome research 28.11 (2018): 1747-1756.
```


## LICENSE

MIT &copy; 2019 Shixiang Wang, Geoffrey Macintyre, Xue-Song Liu
